### ISSUE TITLE
feat: ADR (Architecture Decision Record) Ruleset

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/mdbook-lint#357]
## Summary

Add a new `AdrRuleProvider` to mdbook-lint-rulesets that validates Architecture Decision Records (ADRs) against common formats (Nygard and MADR 4.0.0).

## Motivation

ADRs are widely used in software projects to document architectural decisions. They follow specific formats with required sections, metadata, and cross-referencing conventions. Currently, there's no tooling to lint ADR documents for structural correctness, which leads to:

- Inconsistent ADR formatting across teams
- Missing required sections discovered only during review
- Broken cross-references between ADRs
- Invalid status values or date formats
- Difficulty enforcing organizational ADR standards

An ADR ruleset would complement the existing mdBook ruleset, as ADRs are often included in mdBook documentation.

## Supported Formats

The ruleset should support two primary ADR formats:

### Nygard Format (Michael Nygard - adr-tools)

```markdown
# 1. Use Rust for implementation

Date: 2024-01-15

## Status

Accepted

Supersedes [0001-use-go.md](0001-use-go.md)

## Context

We need to choose a programming language...

## Decision

We will use Rust.

## Consequences

Team will need Rust training.
```

### MADR 4.0.0 (Markdown Any Decision Records)

```markdown
---
status: accepted
date: 2024-01-15
decision-makers:
  - Alice Smith
consulted:
  - Bob Jones
informed:
  - Carol White
---

# Use PostgreSQL for persistence

## Context and Problem Statement

We need to select a database...

## Decision Drivers

* Need ACID compliance
* Team familiarity

## Considered Options

* PostgreSQL
* MySQL
* MongoDB

## Decision Outcome

Chosen option: PostgreSQL, because...

### Consequences

* Good, because mature ecosystem
* Bad, because operational complexity
```

### Structure Rules

| Rule ID | Name | Description | Severity |
|---------|------|-------------|----------|
| ADR001 | `adr-title-format` | H1 title should follow "# NUMBER. TITLE" pattern (Nygard) or be present (MADR) | Error |
| ADR002 | `adr-required-status` | Status section/frontmatter field must exist | Error |
| ADR003 | `adr-required-date` | Date must be present | Error |
| ADR004 | `adr-required-context` | Context/Context and Problem Statement section required | Error |
| ADR005 | `adr-required-decision` | Decision/Decision Outcome section required | Error |
| ADR006 | `adr-required-consequences` | Consequences section required | Warning |

### Validation Rules

| Rule ID | Name | Description | Severity |
|---------|------|-------------|----------|
| ADR007 | `adr-valid-status` | Status must be one of: Proposed, Accepted, Deprecated, Superseded (configurable) | Error |
| ADR008 | `adr-date-format` | Date must be valid ISO 8601 (YYYY-MM-DD) | Error |
| ADR009 | `adr-number-title-match` | ADR number in title must match filename number | Error |
| ADR010 | `adr-filename-convention` | Filename must match `NNNN-slug.md` pattern | Warning |

### Link Rules

| Rule ID | Name | Description | Severity |
|---------|------|-------------|----------|
| ADR011 | `adr-valid-links` | Cross-references to other ADRs must point to existing files | Error |
| ADR012 | `adr-superseded-has-link` | ADRs with Superseded status should have a "superseded by" link | Warning |
| ADR013 | `adr-link-format` | ADR links should follow format: `[NNNN-title](NNNN-slug.md)` | Warning |

### Content Quality Rules

| Rule ID | Name | Description | Severity |
|---------|------|-------------|----------|
| ADR014 | `adr-non-empty-sections` | Required sections should not be empty or contain only placeholder text | Warning |
| ADR015 | `adr-decision-drivers-format` | Decision Drivers (MADR) should be a bullet list | Info |
| ADR016 | `adr-considered-options-format` | Considered Options should list at least 2 options | Info |
| ADR017 | `adr-consequences-structure` | Consequences should distinguish good/bad outcomes (MADR) | Info |

### Repository Rules (Optional - Multi-file Analysis)

| Rule ID | Name | Description | Severity |
|---------|------|-------------|----------|
| ADR018 | `adr-sequential-numbering` | ADR numbers should be sequential without gaps | Info |
| ADR019 | `adr-no-duplicate-numbers` | No duplicate ADR numbers in repository | Error |

## Configuration

```toml
[providers.adr]
enabled = true
format = "auto"  # "auto", "nygard", "madr"

[rules.ADR007]
valid-statuses = ["Proposed", "Accepted", "Deprecated", "Superseded", "Draft"]

[rules.ADR010]
require-zero-padding = true
padding-width = 4

[rules.ADR018]
allow-gaps = false
```

### Format Detection

The ruleset should auto-detect format based on:

1. YAML frontmatter present -> MADR format
2. No frontmatter, "Date:" line after H1 -> Nygard format
3. Configuration override available

### Feature Flag

```toml
[features]
adr = []  # Optional, not in default
```

### Dependencies

Minimal - primarily markdown AST analysis. May optionally depend on:

- `yaml-rust2` or `serde_yaml` for frontmatter parsing (if not already available)

## Prior Art

- [adr-tools](https://github.com/npryce/adr-tools) - Shell scripts for ADR management
- [adr-log](https://github.com/adr/adr-log) - ADR log generator
- [adrs](https://github.com/joshrotenberg/adrs) - Rust ADR toolkit with parsing/validation
- [MADR](https://adr.github.io/madr/) - Markdown Any Decision Records specification

## Benefits

1. **Consistency** - Enforce ADR format standards across projects
2. **Early feedback** - Catch structural issues before review
3. **Integration** - Works with existing mdbook-lint infrastructure
4. **Configurable** - Adapt to organizational standards
5. **Format-agnostic** - Support both major ADR formats

## References

- [ADR GitHub Organization](https://adr.github.io/)
- [Michael Nygard's original article](https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions)
- [MADR 4.0.0 Specification](https://adr.github.io/madr/)
- [JSON-ADR Specification](https://github.com/joshrotenberg/adrs/blob/main/schema/json-adr/v1.json)
